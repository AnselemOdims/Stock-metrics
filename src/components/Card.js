import styled from 'styled-components';
import { BsArrowUp, BsArrowDown, BsArrowRightCircle } from 'react-icons/bs';
import { useNavigate } from 'react-router-dom';
import PropTypes from 'prop-types';
import Img from '../images/stock_card.png';

const Container = styled.div`
  border: solid 1px #f5eedc;
  background-color: #e8e8a6;
  padding: 5px 5px 10px 12px;
  border-radius: 5px;
  background-image: url(${Img});
  background-repeat: no-repeat;
  background-position: bottom right;
  background-size: 55% 23%;
  height: 250px;
  color: #133b5c;
  cursor: pointer;

    > div:nth-of-type(1) {
      display: flex;
      justify-content: space-between;
      align-items: center;
      margin-bottom: 5px;

      > h2 {
        margin: 0;
        font-size: 1rem;
      }
    }

    > span {
      background: rgba(247, 250, 211, 0.637);
      border-radius: 8px;
      padding: 3px 10px;
      font-size: 10px;
    }

    > div:nth-of-type(2) {
      display: flex;
      flex-direction: column;
      margin-top: 10px;
      font-size: 14px;

      span:first-of-type {
        margin-bottom: 10px;
        font-weight: bold;
      }
    }
`;
const Card = ({
  id, ticker, changes, price, changesPercentage, companyName,
}) => {
  const navigate = useNavigate();

  return (
    <Container
      onClick={() => navigate(`/details/${id}`)}
    >
      <div data-testid="company">
        <h2>{companyName}</h2>
        <BsArrowRightCircle size="16px" />
      </div>
      <span>{ticker}</span>
      <div data-testid="card-details">
        <span>
          $
          {price}
        </span>
        <span>
          <BsArrowUp color="green" />
          {changes}
        </span>
        <span>
          <BsArrowDown color="red" />
          {changesPercentage}
        </span>
      </div>
    </Container>
  );
};

Card.propTypes = {
  id: PropTypes.string.isRequired,
  ticker: PropTypes.string.isRequired,
  changes: PropTypes.number.isRequired,
  price: PropTypes.string.isRequired,
  changesPercentage: PropTypes.string.isRequired,
  companyName: PropTypes.string.isRequired,
};
export default Card;
