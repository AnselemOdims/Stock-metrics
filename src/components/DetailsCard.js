import styled from 'styled-components';
import PropTypes from 'prop-types';

const Container = styled.div`
  display: flex;
  flex-direction: column;
`;

const DetailsCard = ({
  children, first, second, third,
}) => (
  <Container>
    <span data-testid="first">{first}</span>
    <span data-testid="second">
      {children?.[0]}
      {second}
    </span>
    <span data-testid="third">
      {children?.[1]}
      {third}
    </span>
  </Container>
);

DetailsCard.propTypes = {
  first: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  second: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  third: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  children: PropTypes.node.isRequired,
};

export default DetailsCard;
