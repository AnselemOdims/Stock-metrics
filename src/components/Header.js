import styled from 'styled-components';
import { BsFillMicFill } from 'react-icons/bs';
import { IoIosSettings, IoIosArrowBack } from 'react-icons/io';
import { NavLink } from 'react-router-dom';
import { motion } from 'framer-motion';
import PropTypes from 'prop-types';

const NavBar = styled.header`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 8px 10px;  

  > div:nth-of-type(1) {
    h3 {
      margin: 10px 0;
      color: #133B5C;
    }
  }
  > div:nth-of-type(2) {
    display: flex;

    > span:nth-of-type(1) {
      margin-right: 15px;
    }
  }
`;

const containerVariants = {
  hidden: {
    y: -100,
  },
  visible: {
    y: 0,
    transition: { duration: 1, stiffness: 500 },
  },
};

const Header = ({ home }) => (
  <motion.div variants={containerVariants} initial="hidden" animate="visible">
    <NavBar>
      {!home && (
        <NavLink to="/">
          <IoIosArrowBack size="20px" color="#133B5C" />
        </NavLink>
      )}
      <div>
        <h3>{home ? 'Stock Metrics' : 'Company Details'}</h3>
      </div>
      <div>
        <span><BsFillMicFill size="18px" color="#133B5C" /></span>
        <span>
          {' '}
          <IoIosSettings size="18px" color="#133B5C" />
        </span>
      </div>
    </NavBar>
  </motion.div>
);

Header.propTypes = {
  home: PropTypes.bool.isRequired,
};

export default Header;
