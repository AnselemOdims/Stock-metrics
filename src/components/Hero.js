import styled from 'styled-components';
import { BsArrowUp, BsArrowDown } from 'react-icons/bs';
import { motion } from 'framer-motion';
import PropTypes from 'prop-types';

import stockImg from '../images/stock.svg';
import { sumPrices, sumPercent, sumChanges } from '../utils/helper';

const Section = styled.section`
  > div {
    display: flex;
    flex-direction: column;
    align-items: flex-end;
    height: 35vh;
    background-image: url(${(props) => (props.image ? 'hello.png' : stockImg)});
    background-repeat: no-repeat;
    background-position: top left;
    background-size: 300px 250px;
    color: #133B5C;
    padding: 10px 20px;

    > h2 {
      font-size: 2rem;
      margin: -20px 0 1px;
    }

    > div {
      display: flex;
      flex-direction: column;
      align-items: center;

      span:first-of-type {
          margin-bottom: 10px;
          font-weight: bold;
        }
    }
    @media (min-width: 768px) {
      background-size: 700px 300px
    }
  }
`;

const Hero = ({ text, stocks }) => (
  <Section>
    <motion.div>
      <h2>{text}</h2>
      <div data-testid="hero-details">
        <span>{`$${sumPrices(stocks)}`}</span>
        <span>
          <BsArrowUp color="green" />
          {`+${sumPercent(stocks)}`}
        </span>
        <span>
          <BsArrowDown color="red" />
          {`-${sumChanges(stocks)}`}
        </span>
      </div>
    </motion.div>
  </Section>
);

Hero.propTypes = {
  text: PropTypes.string.isRequired,
  stocks: PropTypes.arrayOf(PropTypes.shape({
    ticker: PropTypes.string.isRequired,
    changes: PropTypes.number.isRequired,
    price: PropTypes.string.isRequired,
    changesPercentage: PropTypes.string.isRequired,
    companyName: PropTypes.string.isRequired,
  })).isRequired,
};

export default Hero;
