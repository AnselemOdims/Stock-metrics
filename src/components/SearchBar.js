import styled from 'styled-components';
import { FaSearch } from 'react-icons/fa';
import PropTypes from 'prop-types';

const Search = styled.div`
  display: flex;

  > input {
    width: 100%;
    padding: 8px;
    border: solid 2px #e8e8a6;
    font-family: 'Poppins', sans-serif;

    &:focus {
      outline: none;
      border: solid 2px #A1B57D;
    }
  }

  > button {
    justify-self: center;
    width: 10%;
    border: none;
    background: #e8e8a6;
    color: #fff;
  }
`;
const SearchBar = ({ handleSearch, data }) => (
  <Search>
    <input type="search" onChange={handleSearch} value={data} placeholder="Search..." />
    <button type="button">
      <FaSearch color="#fff" />
      .
    </button>
  </Search>
);

SearchBar.propTypes = {
  handleSearch: PropTypes.func.isRequired,
  data: PropTypes.string.isRequired,
};

export default SearchBar;
