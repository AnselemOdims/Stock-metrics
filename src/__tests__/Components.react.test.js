import { render, within } from '@testing-library/react';
import '@testing-library/jest-dom';
import { MemoryRouter } from 'react-router-dom';

import Card from '../components/Card';
import DetailsCard from '../components/DetailsCard';
import Header from '../components/Header';
import Hero from '../components/Hero';
import SearchBar from '../components/SearchBar';

describe('Card', () => {
  test('should render component', () => {
    const { asFragment } = render(
      <Card
        id="AAPL"
        ticker="AAPL"
        changes={1.2}
        price="250"
        changesPercentage="0.23"
        companyName="Apple Inc."
      />,
      { wrapper: MemoryRouter },
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render children correctly', async () => {
    const { container } = render(
      <Card
        id="AAPL"
        ticker="AAPL"
        changes={1.2}
        price="250"
        changesPercentage="0.23"
        companyName="Apple Inc."
      />,
      { wrapper: MemoryRouter },
    );

    const companyContainer = container.querySelector('[data-testid="company"]');
    const cardDetails = container.querySelector('[data-testid="card-details"]');

    expect(await within(container).findByText('AAPL')).toBeTruthy();
    expect(companyContainer).toBeInTheDocument();
    expect(companyContainer.children.length).toBe(2);
    expect(await within(companyContainer).findByText('Apple Inc.')).toBeTruthy();
    expect(cardDetails).toBeInTheDocument();
    expect(cardDetails.children.length).toBe(3);
    expect(await within(cardDetails).findByText('$250')).toBeTruthy();
    expect(await within(cardDetails).findByText(1.2)).toBeTruthy();
    expect(await within(cardDetails).findByText('0.23')).toBeTruthy();
  });
});

describe('Details Card', () => {
  test('should render component correctly', () => {
    const { asFragment } = render(
      <DetailsCard
        first="2020-02-02"
        second="Filling Date: 2020-02-05"
        third="Accepted Date: 2020-02-05"
      >
        <span />
      </DetailsCard>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render children correctly', async () => {
    const { container } = render(
      <DetailsCard
        first="2020-02-02"
        second="Filling Date: 2020-02-05"
        third="Accepted Date: 2020-02-05"
      >
        <span />
      </DetailsCard>,
    );

    const first = container.querySelector('[data-testid="first"]');
    const second = container.querySelector('[data-testid="second"]');
    const third = container.querySelector('[data-testid="third"]');

    expect(await within(container).findByText('2020-02-02')).toBeTruthy();
    expect(await within(container).findByText('Filling Date: 2020-02-05')).toBeTruthy();
    expect(await within(container).findByText('Accepted Date: 2020-02-05')).toBeTruthy();
    expect(first).toBeInTheDocument();
    expect(second).toBeInTheDocument();
    expect(third).toBeInTheDocument();
  });
});

describe('Header', () => {
  test('should render component correctly', () => {
    const { asFragment } = render(<Header home />, { wrapper: MemoryRouter });
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render children correctly', async () => {
    const { container } = render(<Header home />, { wrapper: MemoryRouter });

    const nav = container.querySelector('[data-testid="nav"]');
    const icons = container.querySelector('[data-testid="icons"]');

    expect(nav.children.length).toBe(2);
    expect(icons.children.length).toBe(2);
    expect(await within(nav).findByText('Stock Metrics')).toBeTruthy();
  });
});

describe('Hero', () => {
  test('should render correctly', () => {
    const { asFragment } = render(
      <Hero
        text=""
        stocks={[]}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render children correctly', async () => {
    const { container } = render(<Hero text="Today's Total" stocks={[]} />);

    const hero = container.querySelector('[data-testid="hero-details"]');
    const spans = hero.querySelectorAll('span');

    expect(container.children.length).toBe(1);
    expect(await within(container).findByText('Today\'s Total')).toBeTruthy();
    expect(hero.children.length).toBe(3);
    expect(spans.length).toBe(3);
  });
});

describe('SearchBar', () => {
  test('should render correctly', () => {
    const { asFragment } = render(
      <SearchBar
        handleSearch={() => {}}
        data=""
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render children correctly', async () => {
    const { container } = render(
      <SearchBar
        handleSearch={() => {}}
        data=""
      />,
    );
    expect(container.children.length).toBe(1);
    expect(await within(container).findByRole('button')).toBeTruthy();
  });
});
