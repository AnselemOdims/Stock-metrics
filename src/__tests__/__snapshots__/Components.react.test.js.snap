// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render component 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL iDLfPY"
  >
    <div
      data-testid="company"
    >
      <h2>
        Apple Inc.
      </h2>
      <svg
        fill="currentColor"
        height="16px"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="16px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1 8a7 7 0 1 0 14 0A7 7 0 0 0 1 8zm15 0A8 8 0 1 1 0 8a8 8 0 0 1 16 0zM4.5 7.5a.5.5 0 0 0 0 1h5.793l-2.147 2.146a.5.5 0 0 0 .708.708l3-3a.5.5 0 0 0 0-.708l-3-3a.5.5 0 1 0-.708.708L10.293 7.5H4.5z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <span>
      AAPL
    </span>
    <div
      data-testid="card-details"
    >
      <span>
        $250
      </span>
      <span>
        <svg
          color="green"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          style="color: green;"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 15a.5.5 0 0 0 .5-.5V2.707l3.146 3.147a.5.5 0 0 0 .708-.708l-4-4a.5.5 0 0 0-.708 0l-4 4a.5.5 0 1 0 .708.708L7.5 2.707V14.5a.5.5 0 0 0 .5.5z"
            fill-rule="evenodd"
          />
        </svg>
        1.2
      </span>
      <span>
        <svg
          color="red"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          style="color: red;"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 1a.5.5 0 0 1 .5.5v11.793l3.146-3.147a.5.5 0 0 1 .708.708l-4 4a.5.5 0 0 1-.708 0l-4-4a.5.5 0 0 1 .708-.708L7.5 13.293V1.5A.5.5 0 0 1 8 1z"
            fill-rule="evenodd"
          />
        </svg>
        0.23
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Details Card should render component correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gsDKAQ gFZcdC"
  >
    <span
      data-testid="first"
    >
      2020-02-02
    </span>
    <span
      data-testid="second"
    >
      Filling Date: 2020-02-05
    </span>
    <span
      data-testid="third"
    >
      Accepted Date: 2020-02-05
    </span>
  </div>
</DocumentFragment>
`;

exports[`Header should render component correctly 1`] = `
<DocumentFragment>
  <div
    style="transform: translateY(-100px) translateZ(0);"
  >
    <header
      class="sc-dkPtRN byqvGC"
      data-testid="nav"
    >
      <div>
        <h3>
          Stock Metrics
        </h3>
      </div>
      <div
        data-testid="icons"
      >
        <span>
          <svg
            color="#133B5C"
            fill="currentColor"
            height="18px"
            stroke="currentColor"
            stroke-width="0"
            style="color: rgb(19, 59, 92);"
            viewBox="0 0 16 16"
            width="18px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 3a3 3 0 0 1 6 0v5a3 3 0 0 1-6 0V3z"
            />
            <path
              d="M3.5 6.5A.5.5 0 0 1 4 7v1a4 4 0 0 0 8 0V7a.5.5 0 0 1 1 0v1a5 5 0 0 1-4.5 4.975V15h3a.5.5 0 0 1 0 1h-7a.5.5 0 0 1 0-1h3v-2.025A5 5 0 0 1 3 8V7a.5.5 0 0 1 .5-.5z"
            />
          </svg>
        </span>
        <span>
           
          <svg
            color="#133B5C"
            fill="currentColor"
            height="18px"
            stroke="currentColor"
            stroke-width="0"
            style="color: rgb(19, 59, 92);"
            viewBox="0 0 512 512"
            width="18px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416.3 256c0-21 13.1-38.9 31.7-46.1-4.9-20.5-13-39.7-23.7-57.1-6.4 2.8-13.2 4.3-20.1 4.3-12.6 0-25.2-4.8-34.9-14.4-14.9-14.9-18.2-36.8-10.2-55-17.3-10.7-36.6-18.8-57-23.7C295 82.5 277 95.7 256 95.7S217 82.5 209.9 64c-20.5 4.9-39.7 13-57.1 23.7 8.1 18.1 4.7 40.1-10.2 55-9.6 9.6-22.3 14.4-34.9 14.4-6.9 0-13.7-1.4-20.1-4.3C77 170.3 68.9 189.5 64 210c18.5 7.1 31.7 25 31.7 46.1 0 21-13.1 38.9-31.6 46.1 4.9 20.5 13 39.7 23.7 57.1 6.4-2.8 13.2-4.2 20-4.2 12.6 0 25.2 4.8 34.9 14.4 14.8 14.8 18.2 36.8 10.2 54.9 17.4 10.7 36.7 18.8 57.1 23.7 7.1-18.5 25-31.6 46-31.6s38.9 13.1 46 31.6c20.5-4.9 39.7-13 57.1-23.7-8-18.1-4.6-40 10.2-54.9 9.6-9.6 22.2-14.4 34.9-14.4 6.8 0 13.7 1.4 20 4.2 10.7-17.4 18.8-36.7 23.7-57.1-18.4-7.2-31.6-25.1-31.6-46.2zm-159.4 79.9c-44.3 0-80-35.9-80-80s35.7-80 80-80 80 35.9 80 80-35.7 80-80 80z"
            />
          </svg>
        </span>
      </div>
    </header>
  </div>
</DocumentFragment>
`;

exports[`Hero should render correctly 1`] = `
<DocumentFragment>
  <section
    class="sc-hKwDye fKlLEq"
  >
    <div>
      <h2 />
      <div
        data-testid="hero-details"
      >
        <span>
          $0.00
        </span>
        <span>
          <svg
            color="green"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            style="color: green;"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 15a.5.5 0 0 0 .5-.5V2.707l3.146 3.147a.5.5 0 0 0 .708-.708l-4-4a.5.5 0 0 0-.708 0l-4 4a.5.5 0 1 0 .708.708L7.5 2.707V14.5a.5.5 0 0 0 .5.5z"
              fill-rule="evenodd"
            />
          </svg>
          +0.00
        </span>
        <span>
          <svg
            color="red"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            style="color: red;"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 1a.5.5 0 0 1 .5.5v11.793l3.146-3.147a.5.5 0 0 1 .708.708l-4 4a.5.5 0 0 1-.708 0l-4-4a.5.5 0 0 1 .708-.708L7.5 13.293V1.5A.5.5 0 0 1 8 1z"
              fill-rule="evenodd"
            />
          </svg>
          -0.00
        </span>
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`SearchBar should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-eCImPb lmdkNw"
  >
    <input
      type="search"
      value=""
    />
    <button
      type="button"
    >
      <svg
        color="#fff"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        style="color: rgb(255, 255, 255);"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
        />
      </svg>
      .
    </button>
  </div>
</DocumentFragment>
`;
